src/resolvers/nodeResolver.ts

```diff
@@ -66,19 +82,30 @@ export class NodeResolver implements Resolver {
 		}
 	}
 
-	public needsEntry(filename: string): boolean {
+	public needsEntry(
+		entryPath: ParsedPathPlus,
+		specifiedDockerfilePath?: string,
+	): boolean {
+		// Note: tar files always use forward slash as path separators (regardless
+		// of platform/OS conventions), so the search for '/' instead of `path.sep`
+		// below is not a bug :-)
+		const unparsed = entryPath.unparsed;
 		return (
-			filename === 'package.json' ||
-			filename === 'wscript' ||
-			_.endsWith(filename, '.gyp')
+			!specifiedDockerfilePath &&
+			(unparsed === 'package.json' ||
+				unparsed === 'wscript' ||
+				(!unparsed.includes('/') && entryPath.ext === '.gyp'))
```

Comment:
Two cases:
* If a `specifiedDockeriflePath` was specified, then this resolver (`NodeResolver`) will never match (because of the test: `!specifiedDockerfilePath`). This was previously enforced through a filter over the `NodeResolver. allowSpecifiedDockerfile = false` member variable, which I have deleted for all resolvers.
* Otherwise, it will match `package.json`, `wscript` or `*.gyp` at the root of the project directory tree, as `entryPath.unparsed` is the full path. This root-only rule was previously enforced in `resolveTarStreamOnEntry()` through a test: `if (path.normalize(name).split(path.sep).length === 1)`. That test had been added by a recent-ish pull request:
_fix: Correctly resolve contexts with multiple bundle descriptions_
https://github.com/balena-io-modules/resin-bundle-resolve/pull/32/files

Note that both `entryPath` and `specifiedDockerfilePath` are normalized through `TarUtils.normalizeTarEntry()` before the call to `needsEntry()`, so they won't have leading or trailing slashes or redundant path components.